File: python-engine/patch_routes.py
import re

with open('python-engine/api/routes.py', 'r') as f:
    content = f.read()

ws_import = "from fastapi import APIRouter, HTTPException, Path, WebSocket, WebSocketDisconnect"
content = content.replace("from fastapi import APIRouter, HTTPException, Path", ws_import)

ws_endpoint = """
@router.websocket("/ws/analyze")
async def analyze_websocket(websocket: WebSocket):
    await websocket.accept()
    try:
        loop = asyncio.get_running_loop()
        while True:
            data_str = await websocket.receive_text()
            import json
            req = json.loads(data_str)
            req_id = req.get("req_id")
            payload = req.get("payload", {})
            
            # run analysis
            result = await loop.run_in_executor(None, analyzer.analyze, payload)
            
            await websocket.send_text(json.dumps({"req_id": req_id, "data": result}))
    except WebSocketDisconnect:
        logger.info("WebSocket disconnected")
    except Exception as e:
        logger.error(f"WebSocket error: {e}")
"""

if "analyze_websocket" not in content:
    content += ws_endpoint

with open('python-engine/api/routes.py', 'w') as f:
    f.write(content)

File: patch_routes.js
const fs = require('fs');

let content = fs.readFileSync('python-engine/api/routes.py', 'utf-8');

if (!content.includes('from fastapi import APIRouter, HTTPException, Path, WebSocket, WebSocketDisconnect')) {
    content = content.replace('from fastapi import APIRouter, HTTPException, Path', 'from fastapi import APIRouter, HTTPException, Path, WebSocket, WebSocketDisconnect\nimport json');
}

const wsEndpoint = `
@router.websocket("/ws/analyze")
async def analyze_ws(websocket: WebSocket):
    await websocket.accept()
    try:
        while True:
            data = await websocket.receive_text()
            try:
                payload = json.loads(data)
                correlation_id = payload.get("correlation_id")
                req_data = payload.get("payload", {})
                
                loop = asyncio.get_running_loop()
                result = await loop.run_in_executor(None, analyzer.analyze, req_data)
                
                await websocket.send_text(json.dumps({
                    "correlation_id": correlation_id,
                    "result": result
                }))
            except Exception as e:
                logger.error(f"Analysis WS error: {e}")
                if "correlation_id" in locals():
                    await websocket.send_text(json.dumps({
                        "correlation_id": correlation_id,
                        "error": str(e)
                    }))
    except WebSocketDisconnect:
        pass
    except Exception as e:
        logger.error(f"WebSocket fatal error: {e}")

@router.post("/v1/analyze")
`;

if (!content.includes('@router.websocket("/ws/analyze")')) {
    content = content.replace('@router.post("/v1/analyze")', wsEndpoint);
}

fs.writeFileSync('python-engine/api/routes.py', content);
